[AArch64][SVE] Lower unpredicated loads/stores as fixed LDR/STR with -msve-vector-bits=128.

rj-jesus opened the pull request:
Given the code below:
```cpp
svuint8_t foo(uint8_t *x) {
  return svld1(svptrue_b8(), x);
}
```
When compiled with -msve-vector-bits=128 (or vscale_range(1, 1)), we
currently generate:
```gas
foo:
  ptrue   p0.b
  ld1b    { z0.b }, p0/z, [x0]
  ret
```
Whereas (on little-endian) we could instead be using LDR as follows:
```gas
foo:
  ldr     q0, [x0]
  ret
```

Besides avoiding the predicate dependency, the above form enables
further optimisations such as LDP folds. Likewise for other types and
stores.

I have a patch that enables similar folds for SVE LDR/STR, but since that
causes a fair number of test changes I rather open a separate PR for it.

rj-jesus committed:
Precommit unpredicated loads/stores tests

rj-jesus committed:
[AArch64][SVE] Lower unpredicated loads/stores as LDR/STR with sve-vector-bits=128.

Given the code below:
```cpp
svuint8_t foo(uint8_t *x) {
  return svld1(svptrue_b8(), x);
}
```
When compiled with -msve-vector-bits=128 (or vscale_range(1, 1)), we
currently generate:
```gas
foo:
  ptrue   p0.b
  ld1b    { z0.b }, p0/z, [x0]
  ret
```
Whereas (on little-endian) we could instead be using LDR as follows:
```gas
foo:
  ldr     q0, [x0]
  ret
```

Besides avoiding the predicate dependency, the above form enables
further optimisations such as LDP folds. Likewise for stores.

rj-jesus commented on llvm/test/CodeGen/AArch64/sve-unpred-loads-stores.ll:
```diff
@@ -237,8 +223,8 @@ define <vscale x 16 x i8> @ld_nxv16i8_offset(ptr %0) #0 {
 ; CHECK-128:       // %bb.0:
-; CHECK-128-NEXT:    ptrue p0.b
-; CHECK-128-NEXT:    ld1b { z0.b }, p0/z, [x0, #1, mul vl]
+; CHECK-128-NEXT:    rdvl x8, #1
```
The lowering for the `*_offset` versions could be improved, but if the equivalent ACLE code was compiled directly we wouldn't have `vscale` in the IR, and so would have the natural:
```
ldr	q0, [x0, #16]
```
and
```
str	q0, [x0, #16]
```

paulwalker-arm commented on llvm/test/CodeGen/AArch64/sve-unpred-loads-stores.ll:
```diff
@@ -0,0 +1,483 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -aarch64-sve-vector-bits-max=0   < %s | FileCheck %s --check-prefix=CHECK-VLA
```
Given the general SVE code generation should already be well tested.  Does the `-aarch64-sve-vector-bits-max=0` RUN line offer any new value?

rj-jesus replied:
Thanks, I was initially planning to use it to check the general SVE folds to LDR/STR, but you're right, there's already enough coverage for those. I've removed the `-aarch64-sve-vector-bits-max=0` RUN line and cleaned up the tests a bit.

paulwalker-arm reviewed (commented):
Whilst not unreasonable this PR makes me slightly uneasy because we have considerable code within the code generator that goes the other way (i.e. from fixed-length to scalable vector) for:
* wider than NEON fixed length vectors.
* operations unsupported by NEON
* most all fixed length operations when in StreamingSVE mode. 

I worry how far this could travel because if it's worth removing the predicate for the loads and stores then presumably the same is true for many other instructions and then we could end up in a situation where legalisation and combines are conflicting.

I have the following thoughts/questions:
* Where do the scalable vector operations come from?
    * Is this because the cost model has chosen scalable auto-vectorisation when fixed length would have been better?
        * I'd rather not complicate code generation just because LoopVectorize has made the wrong call.
    * Is this scalable vector ACLE code that wants to benefit from knowing the exact vector length?
* If the predicate is the main concern, does emitting the SVE fill/spill instructions improve performance?
    * Which could be achieved during isel.  
    * If this works then perhaps AArch64LoadStoreOpt could be taught to pair SVE spill/fill instructions when the vector length is known to be 128-bit?
* If this is only the start of taking advantage of knowing the exact vector length then would it be better to have these transformations as a dedicated IR pass?  Then the other optimisers can improve things further and the code generator should just work.

As I say, I'm not against the PR but it would be good to understand the direction of travel early to prevent tying selection in knots.

rj-jesus commented:
Hi @paulwalker-arm, thanks very much for your feedback, I'll address the code changes but first I thought I could reply to your main comments.

> I have the following thoughts/questions:
> 
>     * Where do the scalable vector operations come from?
>       * Is this scalable vector ACLE code that wants to benefit from knowing the exact vector length?

Yes, it's scalable ACLE code compiled with `-msve-vector-bits=128` (see example below).

>     * If the predicate is the main concern, does emitting the SVE fill/spill instructions improve performance?
>       * Which could be achieved during isel.
>       * If this works then perhaps AArch64LoadStoreOpt could be taught to pair SVE spill/fill instructions when the vector length is known to be 128-bit?

I think this should work too. I have a patch that adds patterns for emitting SVE LDR/STR instead of PTRUE LD1/ST1. I can put it up for review if you'd like to try going this route instead. What do you think?

>     * If this is only the start of taking advantage of knowing the exact vector length then would it be better to have these transformations as a dedicated IR pass?  Then the other optimisers can improve things further and the code generator should just work.

Whilst having a dedicated pass could be useful, we are not aiming to pursue this "generally" for all instructions. We care about going from scalable to fixed-length vectors for loads and stores mainly to benefit from LDP/STP folds.

For example, given (https://godbolt.org/z/orbeMTon3):
```cpp
#include <arm_sve.h>

svfloat64_t foo(const double *x) {
  svbool_t pg = svptrue_b64();
  return svld1_f64(pg, x) + svld1_f64(pg, x+svcntd());
}
```
We currently generate:
```gas
foo:
        ptrue   p0.d
        mov     x8, #2
        ld1d    { z0.d }, p0/z, [x0]
        ld1d    { z1.d }, p0/z, [x0, x8, lsl #3]
        fadd    z0.d, z0.d, z1.d
        ret
```
With this patch, we would instead have:
```gas
foo:
	ldp	q0, q1, [x0]
	fadd	v0.2d, v0.2d, v1.2d
	ret
```

Please let me know what you think. I'm happy to try the route via AArch64LoadStoreOpt you suggested if you think that's a better strategy!

paulwalker-arm commented:
@rj-jesus - From the output I see a NEON `fadd`, presumably because an existing combine looks through the insert subvector.  If this is want you want to generate then my AArch64LoadStoreOpt suggestion will not work because it'll bypass the necessary combine.

If you're happy to limit the output changes to just the loads and stores then experimenting with emitting SVE fill/spill instruction and extending AArch64LoadStoreOpt would be nice because it could also optimise real spill/fill code.  This would be especially beneficial across function calls.

Otherwise, given the expected reach as you've described it, I'm happy enough for you to continue down the existing path assuming others agree.  However, given this specifically relates to the ACLE, can the combine be restricted to pre type-legalisation?  I think that'll cover your use case whilst removing the potential conflicts I'm worried about.

rj-jesus commented:
> From the output I see a NEON fadd, presumably because an existing combine looks through the insert subvector. If this is want you want to generate then my AArch64LoadStoreOpt suggestion will not work because it'll bypass the necessary combine.

I think the NEON FADD isn't critical to have - I'd be happy with an SVE FADD too, ideally unpredicated, which I do see emitted with the separate patch for SVE LDR/STR.

> If you're happy to limit the output changes to just the loads and stores then experimenting with emitting SVE fill/spill instruction and extending AArch64LoadStoreOpt would be nice because it could also optimise real spill/fill code. This would be especially beneficial across function calls.

That sounds good, in that case I'll put up a separate patch for the SVE LDR/STR case and separately experiment with the AArch64LoadStoreOpt alternative. In the meantime I'll also address your comments here, I just haven't had the chance to do so today (I'm sorry).

>  However, given this specifically relates to the ACLE, can the combine be restricted to pre type-legalisation? I think that'll cover your use case whilst removing the potential conflicts I'm worried about.

I think that should work, in fact I believe I tried that in an earlier version of the patch and I seem to recall it working.

rj-jesus commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -23550,6 +23550,31 @@ static SDValue combineV3I8LoadExt(LoadSDNode *LD, SelectionDAG &DAG) {
+static SDValue combineVScale1Load(LoadSDNode *LD, SelectionDAG &DAG,
+                                  const AArch64Subtarget *Subtarget) {
+  EVT MemVT = LD->getMemoryVT();
+  if (!Subtarget->isNeonAvailable() || !MemVT.isScalableVector() ||
```
Thanks, I think you're right, I've changed `isNeonAvailable` to `hasNEON`.

It should indeed be restricted to little-endian only, but I believe this is already enforced in `performLOADCombine` (see https://github.com/llvm/llvm-project/blob/90802ec069a88ebce7de370658093ed919480fb8/llvm/lib/Target/AArch64/AArch64ISelLowering.cpp#L23609). Would you rather have a check here in any case?

rj-jesus commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -23550,6 +23550,31 @@ static SDValue combineV3I8LoadExt(LoadSDNode *LD, SelectionDAG &DAG) {
+                                  const AArch64Subtarget *Subtarget) {
+  EVT MemVT = LD->getMemoryVT();
+  if (!Subtarget->isNeonAvailable() || !MemVT.isScalableVector() ||
+      Subtarget->getMaxSVEVectorSizeInBits() != AArch64::SVEBitsPerBlock)
```
Thanks, done.

rj-jesus commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -23550,6 +23550,31 @@ static SDValue combineV3I8LoadExt(LoadSDNode *LD, SelectionDAG &DAG) {
+    return SDValue();
+
+  // Skip unpacked types given their different layouts between Neon and SVE.
+  if (MemVT.getSizeInBits().getKnownMinValue() != AArch64::SVEBitsPerBlock)
```
Thanks, done as well.

You're right, thank you, I've added a check to ignore extending loads.

rj-jesus commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -23845,6 +23873,30 @@ static SDValue combineI8TruncStore(StoreSDNode *ST, SelectionDAG &DAG,
+
+  // Skip unpacked types given their different layouts between Neon and SVE.
+  if (ValueVT.getSizeInBits().getKnownMinValue() != AArch64::SVEBitsPerBlock)
+    return SDValue();
```
Thanks, done.

rj-jesus commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -23550,6 +23550,31 @@ static SDValue combineV3I8LoadExt(LoadSDNode *LD, SelectionDAG &DAG) {
+      LD->getOriginalAlign(), LD->getMemOperand()->getFlags(), LD->getAAInfo());
+  SDValue Insert = convertToScalableVector(DAG, MemVT, NewLoad);
+  SDValue TokenFactor = DAG.getNode(ISD::TokenFactor, DL, MVT::Other,
+                                    {SDValue(cast<SDNode>(NewLoad), 1)});
```
Thank you, you're right, it's not. I've removed it.

paulwalker-arm commented:
> Hi @paulwalker-arm, it seems that doing this in AArch64LoadStoreOpt won't be a walk in the park as by the time we reach the pass, when compiling with `-msve-vector-bits=128`, it's likely the loads/stores will consist of a mix of LDRs+LD1s/STRs+ST1s as seen [here](https://godbolt.org/z/f3eY9jhGs). We could potentially lower the LD1s/ST1s into LDR/STR before combining them into LDP/STP. Do you think this would be preferable to the current implementation?

I see what you mean, the flaw being the instructions don't support a consistent set of addressing modes?  Fair enough, I think for today that means we're better keeping things simple rather than complicating `AArch64LoadStoreOpt`.  Thanks for investigating.

That just leaves the choice over whether to have this in IR or CodeGen.  Have you looked into doing this as an InstCombine?  I'm hoping doing it in IR will be simpler because there's less corner cases t worry about.

rj-jesus commented:
> I see what you mean, the flaw being the instructions don't support a consistent set of addressing modes?

Exactly, this seems to be a (likely unintended) side-effect of compiling for VLS SVE. When compiling for VLS SVE, VL-based offsets are folded to constants early on, which then leads to a mismatch with the addressing modes of SVE loads/stores given they generally expect offsets as multiples of VL ([example](https://godbolt.org/z/5qTEKWG5b)). I think it would be worth "uplifting" these offsets back into multiples of VL, such that they could be folded into immediate-VL-based loads/stores, though I'm not sure where would be the best place to do this. Do you think this is worth trying, probably independently from this PR?

> That just leaves the choice over whether to have this in IR or CodeGen. Have you looked into doing this as an InstCombine?

I haven't tried this yet as I wanted to limit this to the cases we knew we could fold to Neon LDR/STR to promote LDP/STP merging, though I think we could accomplish the same thing with an extra hook that would inspect the vector type and decide accordingly. Is that what you had in mind?

paulwalker-arm commented:
> I haven't tried this yet as I wanted to limit this to the cases we knew we could fold to Neon LDR/STR to promote LDP/STP merging, though I think we could accomplish the same thing with an extra hook that would inspect the vector type and decide accordingly. Is that what you had in mind?

Oh sorry, for some reason I had it in head that the source of the loads/store would all be SVE intrinsics and thus you could extend `instCombineSVELD1`, but that's not always going to be the case.  Please ignore my suggestion.

paulwalker-arm commented:
> I think it would be worth "uplifting" these offsets back into multiples of VL, such that they could be folded into immediate-VL-based loads/stores, though I'm not sure where would be the best place to do this. Do you think this is worth trying, probably independently from this PR?

I've thought the same but it always felt tricky to pull off properly.  The last time I looked at it I wondered whether it could be handling entirely during isel via complex patterns.  I hesitated because I worried it could cause significant noise and make it harder for machine passes that attempt to analyse/merge offsets.  This could be ill founded but with no performance based motivation I figured I'll just ignore it for now.  Feel free to investigate this yourself but for my money I think you'll be better off just pushing this PR through and moving on until analysis shows a genuine need to look deeper.

I'll caveat this by saying despite implementing the feature the whole `-msve-vector-bits` has never sat well with me, although I do appreciate that when maximum performance is required the ideals of scalable vectors becomes secondary :)

rj-jesus commented:
> I've thought the same but it always felt tricky to pull off properly. The last time I looked at it I wondered whether it could be handling entirely during isel via complex patterns. I hesitated because I worried it could cause significant noise and make it harder for machine passes that attempt to analyse/merge offsets. This could be ill founded but with no performance based motivation I figured I'll just ignore it for now. Feel free to investigate this yourself but for my money I think you'll be better off just pushing this PR through and moving on until analysis shows a genuine need to look deeper.
> 
> I'll caveat this by saying despite implementing the feature the whole `-msve-vector-bits` has never sat well with me, although I do appreciate that when maximum performance is required the ideals of scalable vectors becomes secondary :)

Thanks for the suggestion, I'll ask around if there are known cases where missing out on the addressing modes is causing problems. In any case, although that wasn't the main goal, the current patch should also help fix the addressing mode issues for VLS 128.

For what it's worth, I'm not the biggest fan of `-msve-vector-bits`/VLS SVE either. I wish the best performance could _always_ be plain SVE. I suppose `-msve-vector-bits` is just a stop-gap solution until we get there. :)

rj-jesus commented:
I've rebased this to see if it gets through the Windows test that was seemingly failing.

In the meantime, I think we could also implement this as a few new patterns in `AArch64SVEInstrInfo.td`, similar to `unpred_load`, but that lower to LDRQui/LDRQroX instead (and likewise for stores). We would probably need a new predicate to check if `sve-vector-bits=128` for this. Do you think that would be any better @paulwalker-arm?

paulwalker-arm commented:
> I've rebased this to see if it gets through the Windows test that was seemingly failing.
> 
> In the meantime, I think we could also implement this as a few new patterns in `AArch64SVEInstrInfo.td`, similar to `unpred_load`, but that lower to LDRQui/LDRQroX instead (and likewise for stores). We would probably need a new predicate to check if `sve-vector-bits=128` for this. Do you think that would be any better @paulwalker-arm?

Definitely not. Unless absolutely necessary, let's keep SVE VLS separate from isel.  Sorry for the sporadic reviews, I will take another look at the PR tomorrow.

rj-jesus commented:
> Definitely not. Unless absolutely necessary, let's keep SVE VLS separate from isel. Sorry for the sporadic reviews, I will take another look at the PR tomorrow.

Sounds good, thank you for the quick sanity check!

rj-jesus committed:
Address comments and rebase patch

rj-jesus commented:
Hi @paulwalker-arm, I'm very sorry for taking so long to update this. Initially, I was waiting on #129732 to get merged to propose the alternative version that combines SVE LDR/STR into LDP/STP directly, but then that PR took a bit longer than expected and I had to put this on hold.

I believe I've addressed most of your feedback. I also have a tentative patch implementing the SVE LDR/STR pairing you initially suggested. Would you prefer I open a separate PR for this, or should I include it in this patch so we can decide on the approach to go for?

Thanks very much!

rj-jesus commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -23589,6 +23589,28 @@ static SDValue combineV3I8LoadExt(LoadSDNode *LD, SelectionDAG &DAG) {
+                                  TargetLowering::DAGCombinerInfo &DCI,
+                                  const AArch64Subtarget *Subtarget) {
+  EVT MemVT = LD->getMemoryVT();
+  if (!DCI.isBeforeLegalize() || !Subtarget->hasNEON() ||
```
I believe this was already checked by the caller (`performLOADCombine()`), but in any case I've added the explicit check here.

rj-jesus commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -23589,6 +23589,28 @@ static SDValue combineV3I8LoadExt(LoadSDNode *LD, SelectionDAG &DAG) {
+      NewVT, DL, LD->getChain(), LD->getBasePtr(), LD->getPointerInfo(),
+      LD->getOriginalAlign(), LD->getMemOperand()->getFlags(), LD->getAAInfo());
+  SDValue Insert = convertToScalableVector(DAG, MemVT, NewLoad);
+  return DAG.getMergeValues({Insert, SDValue(cast<SDNode>(NewLoad), 1)}, DL);
```
Thanks, done.

rj-jesus commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -23589,6 +23589,28 @@ static SDValue combineV3I8LoadExt(LoadSDNode *LD, SelectionDAG &DAG) {
+
+  SDLoc DL(LD);
+  MVT NewVT = MVT::getVectorVT(MemVT.getVectorElementType().getSimpleVT(),
+                               MemVT.getVectorMinNumElements());
```
Thanks, done.

rj-jesus commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -23884,6 +23909,29 @@ static SDValue combineI8TruncStore(StoreSDNode *ST, SelectionDAG &DAG,
+                                   const AArch64Subtarget *Subtarget) {
+  SDValue Value = ST->getValue();
+  EVT ValueVT = Value.getValueType();
+  if (ST->isVolatile() || !Subtarget->isLittleEndian() ||
```
You're right, I don't think it does. I've removed this.

paulwalker-arm commented:
> I believe I've addressed most of your feedback. I also have a tentative patch implementing the SVE LDR/STR pairing you initially suggested. Would you prefer I open a separate PR for this, or should I include it in this patch so we can decide on the approach to go for?

Please use a separate PR for the SVE LDR/STR pairing approach. Doing this means this PR can remain in reserve if for some reason we decide to reverse the current direction of travel.

rj-jesus commented:
> Please use a separate PR for the SVE LDR/STR pairing approach. Doing this means this PR can remain in reserve if for some reason we decide to reverse the current direction of travel.

Thanks very much, I've just opened #134068 to attempt the SVE fill/spill pairing route.

rj-jesus commented:
I'm closing this since we merged an alternative approach in: https://github.com/llvm/llvm-project/commit/1df4af6c.
